content/blog/2019-11-07-immix/index.md

```diff
@@ -0,0 +1,210 @@
+Immix's performance is consistently among the best. 
+And you can also see there is a fairly large gap between traditional tracing algorithm and generational composite.
+
+The authors also evaluated the performance sensitivity of immix. 
+
+![006.png](./006.png)
+
+The table above shows the performance of immix when certain features are disabled, 
+or parameters are changed. 
+Here I want to focus on the algorithm features.
+The four columns under algorithm features represent performance for mark-region collectors that provide only block marking (Block);
+block and line marking with overflow allocation, 
+but without defragmentation (Line); 
+everything but headroom (No HR); 
+everything but overflow allocation. 
+It is interesting to me that disabling headroom affects the performance a lot while disabling overflow allocation doesn't matter too much.
+
+# Conclusion and Thoughts
+
+This paper introduces mark-region, a new family of garbage collectors, 
+and opportunistic evacuation, 
+a lightweight defragmentation mechanism. 
+It combines them in immix and shows that it matches or beats existing canonical garbage collectors by presenting a full-scale evaluation and detailed analysis.
+
+- I think this paper puts a lot of effort into details of immix's implementation, which is good. However, I see a lot of optimization applied to the original design, for example, adding overflow allocator that always uses free blocks, adding headroom that is reversed for evacuation only. And it seems some of them matter a lot to the performance (by looking at the last table I show), it seems questionable to me that this algorithm would still perform better if we strip these optimizations off. Or maybe all production collectors tweak a lot?
+- This paper doesn't talk about pause times. Maybe the reason is that all the algorithms here stop-the-world, and the techniques for improving pause times such as incremental collection or concurrent collection are orthogonal to the topic.
+
+# Influence
+
+I wanted to find out any existing industrial instance of this algorithm, 
+so I did a little survey on several popular languages' compilers to get a sense of the state of modern garbage collection.
+
+- *Python* (CPython) and [*PHP*](https://www.php.net/manual/en/features.gc.php). They are using reference counting plus some cycle detection mechanism
+- [*Swift*](https://docs.swift.org/swift-book/LanguageGuide/AutomaticReferenceCounting.html). It also uses reference counting, but no cycle detection. Programmers are supposed to leverage weak references correctly to get rid of cycles by themselves. Wow.
+- [*C#*](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/fundamentals?redirectedfrom=MSDN). Concurrent generational mark-compact.
+- *Javascript* ([V8](https://v8.dev/blog/trash-talk)). Parallel concurrent incremental generational semi-space+mark-compact, with idle time GC.
```

Comment:
Capitalize JavaScript.

Reply:
Also, FWIW, I mentioned the [WebKit JavaScript engine's collector](https://webkit.org/blog/7122/introducing-riptide-webkits-retreating-wavefront-concurrent-garbage-collector/) the other day. I don't see any discussion of regions in there.